transmission.go

```diff
@@ -119,99 +119,41 @@ func (t *txTestClient) Add(ev *Event) {
+	batches          map[string][]*Event
 	httpClient       *http.Client
 	blockOnResponses bool
-	numEncoded       int
+	// numEncoded       int
 
 	// allows manipulation of the value of "now" for testing
 	testNower   nower
 	testBlocker *sync.WaitGroup
 }
 
-func (b *batch) Add(ev interface{}) {
+// batch is a collection of events that will all be POSTed as one HTTP call
+// type batch []*Event
+
+func (b *batchAgg) Add(ev interface{}) {
 	// from muster godoc: "The Batch does not need to be safe for concurrent
 	// access; synchronization will be handled by the Client."
-	if b.events == nil {
-		b.events = map[string][]*Event{}
+	if b.batches == nil {
+		b.batches = map[string][]*Event{}
 	}
 	e := ev.(*Event)
-	if _, ok := b.events[e.Dataset]; !ok {
-		// make a new event list if we haven't yet seen this dataset in this batch
-		b.events[e.Dataset] = make([]*Event, 0, 1)
-	}
-	b.events[e.Dataset] = append(b.events[e.Dataset], e)
-}
-
-// Provides finer-tuned control over handling serialization errors by returning
-// individual errors down the responses channel. This lets us successfully deal
-// with a batch that has a mix of OK and broken events.
-func (b *batch) MarshalJSON() ([]byte, error) {
-	buf := bytes.Buffer{}
-	buf.WriteByte('{')
-
-	keys := make([]string, 0, len(b.events))
-	for datasetName := range b.events {
-		keys = append(keys, datasetName)
-	}
-	sort.Strings(keys) // parity with encoding/json's mapEncoder
-
-	var mapCt int
-	var arrCt int
-	for _, datasetName := range keys {
-		kBytes, err := json.Marshal(datasetName) // lean on encoding/json's stringEncoder
-		if err != nil {
-			// we couldn't encode the dataset name; return err for all of the dataset's events
-			for _, ev := range b.events[datasetName] {
-				b.enqueueResponse(Response{
-					Err:      err,
-					Metadata: ev.Metadata,
-				})
-			}
-			continue
-		}
-		if mapCt > 0 { // if we're on our 2nd+ dataset with encoded events
-			buf.WriteByte(',')
-		}
-		mapCt++
-		buf.Write(kBytes)
-		buf.WriteByte(':')
-		buf.WriteByte('[')
-
-		arrCt = 0
-		for i, ev := range b.events[datasetName] {
-			eBytes, err := json.Marshal(ev)
-			if err != nil {
-				b.enqueueResponse(Response{
-					Err:      err,
-					Metadata: ev.Metadata,
-				})
-				// nil out the invalid Event so we can line up sent Events with server
-				// responses if needed. don't delete to preserve slice length.
-				b.events[datasetName][i] = nil
-				continue
-			}
-			if arrCt > 0 {
-				buf.WriteByte(',')
-			}
-			arrCt++
-			b.numEncoded++
-			buf.Write(eBytes)
-		}
-		buf.WriteByte(']')
+	// collect separate buckets of events to send based on the trio of api/wk/ds
+	// if all three of those match it's safe to send all the events in one batch
+	key := fmt.Sprintf("%s_%s_%s", e.APIHost, e.WriteKey, e.Dataset)
+	if _, ok := b.batches[key]; !ok {
+		// make a new event list if we haven't yet seen this key in this batch
+		b.batches[key] = make([]*Event, 0, 1)
```

Comment:
You know what, this whole check is actually unnecessary. `append()` will do the right thing if the first argument is a `nil`, which the `map` will return.